### PR TITLE
Support Comma Separated Cookies

### LINKED ISSUE #1434: cookies: support multi-cookie, comma-separated header variant
At the moment, we only support strictly compliant cookies, where each cookie is in a separate header. The justification for this we give in the module docstring is that "violations of RFC6265 that makes the parsing problem indeterminate are much more common than genuine occurences of the multi-cookie variants". However, a recent sample from Netograph shows old-style cookies in ~9% of sites, with many high-profile sources like DoubleClick. There's no way around it - we simply must correctly parse this variant.

### PR DESCRIPTION
**Pull request by dufferzafar**
Fixes #1434

Todo:

- [x] Rebased with master
- [x] The `_read_pairs` change should only affect `SetCookie` and not `Cookie`
- [x] Update `format_set_cookie_header` to take in a list of `SetCookie`s & return the comma separated list.
- [x] Add new tests for multi-cookies

### CODE REVIEW

File: netlib/http/cookies.py [hunk 1]
```diff
@@ -161,10 +175,11 @@ def _parse_set_cookie_pairs(s):
...
     ret = []
     for header in headers:
-        v = parse_set_cookie_header(header)
-        if v:
-            name, value, attrs = v
-            ret.append((name, SetCookie(value, attrs)))
+        cookies = parse_set_cookie_header(header)
+        if cookies:
+            for cookie in cookies:
+                name, value, attrs = cookie
```
Comment (Kriechi) [inline netlib/http/cookies.py]: I think you can directly use this as loop args: `for name, value, attrs in cookies:`

File: netlib/http/cookies.py [hunk 2]
```diff
@@ -187,13 +202,17 @@ def parse_set_cookie_header(line):
...
-    pairs = _parse_set_cookie_pairs(line)
-    if pairs:
-        return pairs[0][0], pairs[0][1], CookieAttrs(tuple(x) for x in pairs[1:])
+    cookies = [
+        (pairs[0][0], pairs[0][1], CookieAttrs(tuple(x) for x in pairs[1:]))
+        for pairs in _parse_set_cookie_pairs(line) if pairs
+    ]
+
+    if cookies:
+        return cookies
```
Comment (Kriechi) [inline netlib/http/cookies.py]: This implicitly return None - which a neat way to do it - but I fear a bit too much 😄 
I think we should make it explicit in that case - what do you think?

**Comment by Kriechi** [comment]
Maybe the py.test parametrize feature would be a good fit to cleanup the cookie tests?
http://docs.pytest.org/en/latest/parametrize.html

**Comment by dufferzafar** [comment]
@cortesi @resam I understand that we should separate cookies based on comma in the `SetCookie` header.

But what should happen if there is a comma in the `Cookie` header.

For eg: in `Cookie: name=value, key=value"

**Comment by cortesi** [comment]
The RFC says that cookie values can't contain commas, so let's assume that's correct for now. So for a first pass, we can assume that the only place we need to deal with comma ambiguity is in the expires attribute. The caveat here is that browsers might do something different, and if they do, we should support it. We can get to that later. 

I'm super excited about this PR because I have a very real use for this. If you could rebase and push  ahead with the fixes you've outlined, I'll test on the Netograph corpus and merge.

**Comment by mhils** [comment]
@cortesi @dufferzafar: What's the status here? 😃

**Comment by dufferzafar** [comment]
I fixed all the current tests this week. Now I need to add some new tests specific to comma separated cookies, and this will be done.

**Comment by mhils** [comment]
Sounds great, thanks for the update! :smiley:

**Comment by dufferzafar** [comment]
New tests added! Ready for review.

**Pull request merged**